### PR TITLE
Replace 968

### LINKED ISSUE #955: Commands with arguments in menus
This issue concerns how the define-command macro (and similar command defining macros created by define-application-frame) handle commands that go into a menu (i.e. :menu t is specified).  clim:Define-command expands into clim-internals::%define-command which then expands into a variety of function and method definitions as well as a call to clim:add-command-to-command-table.  This call has a keyword argument :menu-command whose value is a list of the command name followed by as many occurrences of clim:*unsupplied-argument-marker* as there are arguments to the command.  For example (from an application of mine) a command with one argument
```
(define-aplan-command (com-load-model :name t :menu t)
    ((pathname 'clim:pathname)) 
...)
```
expands into:
```
(clim:add-command-to-command-table 'com-load-model 'aplan :name t :menu 't
                                    :keystroke 'nil :errorp nil :menu-command
                                    (list 'com-load-model
                                          clim:*unsupplied-argument-marker*))
```
the value of clim:*unsupplied-argument-marker* is CLIM-INTERNALS::%UNSUPPLIED-ARGUMENT-MARKER%.

This has two consequences one annoying and one serious:
1) If the menu item is clicked on rather than prompting for the arguments, it tells you you're missing arguments.
2) If the menu is in an application frame with a pointer-documentation pane then just moving the mouse over the menu item will cause an error.  The reason for this is that the pointer documentation pane will attempt to present the command, which in turn attempts to present the arguments.  But the present method for the argument is passed the symbol CLIM-INTERNALS::%UNSUPPLIED-ARGUMENT-MARKER% as the object and the presentation type of the argument (in my example above clim:pathname).   In this example, this means that the presentation method for pathnames is invoked but is passed a symbol as the object.  This can easily cause an error.

None of this occurs in the existing implementation of CLIM (e.g. the ACL one).  In the existing implementations, the call to add-command-to-command-table would supply just (list 'com-load-model) as the :menu-command argument.  As a result the pointer-documentation pane just displays the command name.  The command parser in the ACL implementation prompts for the remaining arguments.  Making the macro expansion do what the ACL implementation does would be easy and would prevent an error from being signaled, solving issue 2, but I don't understand the command parser well enough to solve issue 1.  Anybody have some guidance to offer?

### PR DESCRIPTION
**Pull request by dkochmanski**
fixes #955
fixes #967 
fixes #968

### CODE REVIEW

File: Core/clim-core/commands/processor.lisp [hunk 1]
```diff
@@ -361,14 +361,14 @@
...
-                    (position *unsupplied-argument-marker* command)))
-          (t (values command type)))))
+  (if-let ((command (funcall *command-parser* command-table stream)))
+    (values (if (partial-command-p command)
+                (funcall *partial-command-parser*
+                         command-table stream command
+                         (position *unsupplied-argument-marker* command))
+                command)
+            type)
+    (simple-parse-error "Empty command")))
```
Comment (scymtym) [inline Core/clim-core/commands/processor.lisp]: Can we use the new `accept.smoke` test to ensure this requirement (that is returning two values)?

File: Core/clim-core/commands/processor.lisp [hunk 2]
```diff
@@ -361,14 +361,31 @@
...
+                          command-table stream command
+                          (position *unsupplied-argument-marker* command))
+                 command)))
+      (with-input-context (type)
+          (object type)
+          (if-let ((command (funcall *command-parser* command-table stream)))
+            (values (maybe-replace-input (handle-command command)) type)
+            (simple-parse-error "Empty command"))
+        (command
+         (values (maybe-replace-input (handle-command object)) type))))))
```
Comment (scymtym) [inline Core/clim-core/commands/processor.lisp:393]: Can we use the `accept.smoke` test by feeding a suitable event to the fake extended input stream?

File: Core/clim-core/commands/processor.lisp [hunk 3]
```diff
@@ -38,9 +38,14 @@
...
     (with-delimiter-gestures (*command-argument-delimiters* :override t)
-      (setq command-args (funcall (parser (gethash command-name
-                                                   *command-parser-table*))
-                                  stream)))
+      ;; The null context is a barrier which prevents other commands from
+      ;; being selected during reading arguments. -- jd 2020-09-03
+      (with-input-context ('null :override t) ()
+          (setq command-args (funcall (parser (gethash command-name
+                                                       *command-parser-table*))
+                                      stream))
```
Comment (scymtym) [inline Core/clim-core/commands/processor.lisp:46]: Will this also prevent commands from being selected as arguments? If not, the comment could be clarified a bit.

File: Core/clim-core/commands/processor.lisp [hunk 4]
```diff
@@ -38,9 +38,14 @@
...
+        ;; Impossible clause.
+        (t (error "U Can't Touch This"))))
```
Comment (scymtym) [inline Core/clim-core/commands/processor.lisp:48]: Why is this needed? If it is actually reachable (or in any case, really), the error message should provide information and explain the situation.

**Review by scymtym — CHANGES_REQUESTED** [review]
Questions and suggestions in commit comments.

File: Core/clim-core/commands/processor.lisp (hunk 4 above)
Comment (dkochmanski) [reply Core/clim-core/commands/processor.lisp:48]: It is not reachable, if it is reached then we have a bug in input context processing. it is similar to "You lose" errors in other parts of the code.

File: Core/clim-core/commands/processor.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-core/commands/processor.lisp:46]: if an argument type is command, then it will be put as the context by the command parser. In other words barrier cancels all input contexts from above.

File: Core/clim-core/commands/processor.lisp (hunk 2 above)
Comment (dkochmanski) [reply Core/clim-core/commands/processor.lisp:393]: do you mean something like that?


```
(test accept.smoke
  (mapc
   (lambda (case)
     (destructuring-bind (input accept-type expected-value expected-type) case
       (flet ((test-stream (stream)
                (multiple-value-bind (value type)
                    (accept accept-type :stream stream :prompt nil)
                  (is (equal expected-value value)
                      "~@<ACCEPTing ~S from a ~S with type ~S resulted ~
                       in value ~S, not ~S~@:>"
                      input (type-of stream) accept-type value expected-value)
                  (is (equal expected-value value)
                      "~@<ACCEPTing ~S from a ~S with type ~S resulted ~
                       in type ~S, not ~S~@:>"
                      input (type-of stream) accept-type type expected-type)
                  (is (not (null type))
                      "~@<ACCEPTing ~S from a ~S with type ~S resulted ~
                       in type ~s, which is not a valid."))))
         ;; Test with extended input stream.
         (let ((stream (make-instance 'test-stream)))
           (map nil (alexandria:curry #'climi::stream-append-gesture stream)
                input)
           (climi::stream-append-gesture stream #\Return)
           (test-stream stream))
         ;; Test with string stream.
         (let* ((string (format nil "~A~C" input #\Return))
                (stream (make-string-input-stream string)))
           (test-stream stream)))))
   '(("1"       integer            1         integer)
     ("1,2,3,4" (sequence integer) (1 2 3 4) (sequence integer)))))
```

or do you mean, that we should define a command table and a command, then try to parse it? that won't work for partial command parsers (and that's where we were bitten), because they invoke accepting-values dialog.

Comment (scymtym) [reply Core/clim-core/commands/processor.lisp:393]: If I understand the changed code correctly, we would have to add cases that replace feeding `"1,2,3,4"` to the stream by feeding a simulated presentation-click event to it.

Check the second return value can be done in all cases it seems.

File: Core/clim-core/commands/processor.lisp (hunk 3 above)
Comment (scymtym) [reply Core/clim-core/commands/processor.lisp:46]: So maybe "The null context is a barrier which presents other commands from being selected and cancelling the current command while reading arguments. Selecting commands as arguments to the current command still works."?

File: Core/clim-core/commands/processor.lisp (hunk 4 above)
Comment (scymtym) [reply Core/clim-core/commands/processor.lisp:48]: Maybe "An error an error in McCLIM caused code to be executed that should never be reached. Please report this as a bug."?

File: Core/clim-core/commands/processor.lisp (hunk 1 above)
Comment (scymtym) [reply Core/clim-core/commands/processor.lisp]: Too difficult with the current test infrastructure.

File: Core/clim-core/commands/processor.lisp (hunk 2 above)
Comment (scymtym) [reply Core/clim-core/commands/processor.lisp:393]: Too difficult with the current test infrastructure.

File: Core/clim-core/commands/processor.lisp (hunk 4 above)
Comment (dkochmanski) [reply Core/clim-core/commands/processor.lisp]: removed the clause whatsoever.

File: Core/clim-core/commands/processor.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-core/commands/processor.lisp]: updated the comment to

```
(with-delimiter-gestures (*command-argument-delimiters* :override t)
      ;; The null context is a barrier which prevents other commands from
      ;; being selected during reading arguments and cancelling the current
      ;; command while reading arguments. When the argument's type is command
      ;; then this context will be estabilished by the parser and it will be
      ;; possible to select commands. -- jd 2020-09-03
      (with-input-context ('null :override t) ()
          (setq command-args (funcall (parser (gethash command-name
                                                       *command-parser-table*))
                                      stream))))
```

Comment (dkochmanski) [reply Core/clim-core/commands/processor.lisp]: updated again.

**Review by scymtym — APPROVED** [review]

**Comment by scymtym** [comment]
Looks good to me.
